src/odemis/driver/smaract.py

```diff
@@ -3084,3 +3085,1139 @@ def SA_CTL_Move(self, handle, ch, pos_pm, _):
+        self.DisableEventNotification(SA_SIDLL.SA_PS_STABLE_STATE_CHANGED_EVENT)
+
+    def SaveConfiguration(self):
+        """
+        Saves the configuration parameters (cf LoadConfiguration) and internal parameters of the
+        referencing procedure. This function is blocking.
+        """
+        self.SetProperty_i32(SA_SIDLL.SA_PS_AF_ADJUSTMENT_RESULT_SAVE_PROP, SA_SIDLL.SA_SI_ENABLED)
+        # There is no documentation on loading/saving the configuration in the manual, even though
+        # the functions are available in the header file. It is not clear how much time saving the
+        # parameters takes. It can be assumed that it's fast --> wait 1 s.
+        time.sleep(1)
+
+    def ValidateChannels(self):
+        """
+        Validate the channels. This function is blocking.
+        This is a quick adjustment routine which needs to be performed after loading the configuration.
+        """
+        # There is also a channel validation progress event, but it's triggered too early. We need
+        # to wait until the state becomes stable again before continuing.
+        self.EnableEventNotification(SA_SIDLL.SA_PS_STABLE_STATE_CHANGED_EVENT)
+        self.SetProperty_i32(SA_SIDLL.SA_PS_AF_CHANNEL_VALIDATION_STATE_PROP, SA_SIDLL.SA_SI_ENABLED)
+        self.WaitForEvent(SA_SIDLL.SA_SI_ENABLED)
+        self.DisableEventNotification(SA_SIDLL.SA_PS_STABLE_STATE_CHANGED_EVENT)
+
+    def EnableChannel(self, channel):
+        """
+        Enable channel. This function is blocking.
+        channel (int): channel number
+        """
+        self.SetProperty_i32(SA_SIDLL.SA_PS_CH_ENABLED_PROP, SA_SIDLL.SA_SI_ENABLED, idx0=channel)
+        # Channel enabled event not triggered if it was already enabled
+        # --> poll SA_PS_CH_ENABLED_PROP attribute instead
+        while self.GetProperty_i32(SA_SIDLL.SA_PS_CH_ENABLED_PROP, idx0=channel) != SA_SIDLL.SA_SI_ENABLED:
+            time.sleep(0.1)
+
+    def EnableEventNotification(self, event_type):
+        """
+        Enables notifications for an event type.
+        event_type: (int) SA_SIDLL event type
+        """
+        encoded_key = SA_SIDLL.SA_SI_EVENT_NOTIFICATION_ENABLED_PROP << 16 | event_type
+        self.core.SA_SI_SetProperty_i32(self._id, c_uint32(encoded_key), c_int32(SA_SIDLL.SA_SI_ENABLED))
+
+    def DisableEventNotification(self, event_type):
+        """
+        Disables notifications for an event type.
+        event_type: (int) SA_SIDLL event type
+        """
+        encoded_key = SA_SIDLL.SA_SI_EVENT_NOTIFICATION_ENABLED_PROP << 16 | event_type
+        self.core.SA_SI_SetProperty_i32(self._id, c_uint32(encoded_key), c_int32(SA_SIDLL.SA_SI_DISABLED))
+
+    def WaitForEvent(self, end_state, timeout=float("inf")):
+        """
+        Blocks until event is triggered or timeout.
+        end_state (int): state of event variable to wait for
+        timeout (float): maximum time to wait in s. If inf, it will wait forever.
+        """
```

Comment:
I think that you're making it too fancy. You should have a function which just wraps the underlying library function. So it would return whichever event.
*In addition* you could have a second function `_wait_for_event(state)` which actually does what this function does.
Please note that currently, the timeout doesn't work as expected. The proper way to do that is to compute at the beginning the maximum time `tend` this function can wait until. Based on that value and the current time, the timeout to WaitForEvent() should be adjusted.